### PR TITLE
Side menu

### LINKED ISSUE #19: icon / side menu
Icon to open a side menu, with search field (triggers the search result panel) and history of searches and / or selected gifs.

### LINKED ISSUE #27: Create Tests for V2
A few tests need to be updated while some others need to be created to get better coverage

### PR DESCRIPTION
**Pull request by Carglass**
Ticket: https://github.com/giflens/GifLens/projects/1#card-22245372 <br>
Issue: fixes #19, fixes #27 

## Changes

- added a side menu (view Container)
- two views are accessible, history and favorites (only history is concerned by this PR)
- searching a gif and selecting adds it to the history
- clicking on the link in the history adds the gif as a GIFLENS tag ( temporary, might be replace by an icon action only)

## Remaining [WIP]

- [X] Limit to 15 the number of entries in the history 
- [x] Allow to delete an entry
- [x] Refactor search to have the add Gif action factorized
- [x] Adapt old search tests
- [x] Create test for all new features

## How To Test

- Run the extension, and search a GIF
- confirm that the title of the GIF appears in the history once one is selected

## Applicable Research Resources

- https://code.visualstudio.com/api/extension-guides/tree-view
- https://developers.giphy.com/docs/
- https://github.blog/2013-05-14-closing-issues-via-pull-requests/

### CODE REVIEW

**Comment by Carglass** [comment]
#18

**Comment by Carglass** [comment]
I removed the tests from this PR, they could be related to the necessary reorganization of code. I want this PR to be merged so that we can work separately ( this PR is a requirement for nearly all other 2.0 issues)

File: src/search.ts [hunk 1]
```diff
@@ -34,16 +35,17 @@ export const webviewHtml: (
         var arr = [...searchImages];
 
         // event handler that sends back to the extension the url of the clicked gif
-        const sendUrl = event => {
-          const url = event.target.getAttribute('src');
+        const sendGif = event => {
+					const url = event.target.getAttribute('src');
+					const label = event.target.getAttribute('alt');
```
Comment (fallanic) [inline src/search.ts]: weird indentation here

File: src/search.ts [hunk 2]
```diff
@@ -76,45 +78,52 @@ const errorHtml = `<!DOCTYPE html>
 /**
  * handler for the search TextEditorCommand
  * @param  {vscode.TextEditor} editor the active vsCode editor at the time the command is run
- * @param  {vscode.ExtensionContext} context the context of the extension
+ * @param  {vscode.ExtensionContext} state the context of the extension
```
Comment (fallanic) [inline src/search.ts]: wrong typing here

File: src/search.ts [hunk 3]
```diff
@@ -229,6 +195,22 @@ export const getChosenGifUrl: (
...
+						message.text.label,
+						message.text.url
+					);
+					const nextHistory =
+						// TODO: the 15 could be a setting
+						prevHistory && prevHistory.length < 15
+							? [newEntry].concat(prevHistory)
+							: prevHistory
+							? [newEntry].concat(prevHistory.slice(0, -1))
+							: [newEntry];
```
Comment (fallanic) [inline src/search.ts]: not high priority / critical but linter usually forbif nesting ternary operators.
Maybe we could just do 
```
const nextHistory = [newEntry].concat(prevHistory || []);
if(nextHistory.length >=15){
nextHistory.pop();
}
```

File: src/test/addGif.test.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import {
+	getLanguageCommentEnd,
+	getLanguageCommentStart,
+	addGifLensTagToEditor,
+	// searchTask,
+} from '../addGif';
+
+suite('AddGif creates a GIFLENS tag', function() {
+	test('given a correct language identifier, getLanguageCommentStart should return the correct comment start', function() {
```
Comment (fallanic) [inline src/test/addGif.test.ts]: You get a heart emoji each time for testing my code ❤️

File: src/test/addGif.test.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		const commentStart = getLanguageCommentStart('c');
+		assert(commentStart === '/*');
+	});
+
+	test('given a correct language identifier, getLanguageCommentEnd should return the correct comment end', function() {
```
Comment (fallanic) [inline src/test/addGif.test.ts]: ❤️

File: src/test/deleteGif.test.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+			const historyTreeView = new HistoryProvider(state);
+			const waiter = await state.update('history', [gif]);
+			assert(state.get('history'));
+			// then deletes it
+			deleteGifFromHistory(gif, state, historyTreeView);
+			const final: HistoryEntry[] | undefined = state.get('history');
+			if (final) {
+				assert(final.length === 0);
+			} else {
+				assert(false);
```
Comment (fallanic) [inline src/test/deleteGif.test.ts]: this is great 👍 , let's just add a comment to clarify

File: src/test/search.test.ts [hunk 7]
```diff
@@ -3,30 +3,24 @@ import * as vscode from 'vscode';
...
-		const createdString = createImages(['1', '2', '3']);
+		const createdString = createImages([
+			{ url: '1', label: '4' },
+			{ url: '2', label: '5' },
+			{ url: '3', label: '6' },
+		]);
 		assert(
 			createdString ===
-				'<img class="search-img" style="cursor: pointer;" src="1" /><img class="search-img" style="cursor: pointer;" src="2" /><img class="search-img" style="cursor: pointer;" src="3" />'
+				'<img class="search-img" style="cursor: pointer;" src="1" alt="4" /><img class="search-img" style="cursor: pointer;" src="2" alt="5" /><img class="search-img" style="cursor: pointer;" src="3" alt="6" />'
```
Comment (fallanic) [inline src/test/search.test.ts]: wish we were using Jest snapshots 😇

**Comment by fallanic** [comment]
Also another remark : we should disable favorites in the UI until we actually add the feature

File: src/search.ts (hunk 1 above)
Comment (Carglass) [reply src/search.ts]: I dont know if this is because of the huge string, but the indentation looks ok on my computer :/

**Pull request merged**